File: CS/GanttRestrictions/Form1.cs
using DevExpress.XtraScheduler;
using System;
using System.Data;
using System.Data.SqlClient;
using System.Windows.Forms;

namespace GanttRestrictions
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            // TODO: This line of code loads data into the 'gantTest01DataSet.TaskDependencies' table. You can move, or remove it, as needed.
            this.taskDependenciesTableAdapter.Fill(this.gantTest01DataSet.TaskDependencies);
            // TODO: This line of code loads data into the 'gantTest01DataSet.Resources' table. You can move, or remove it, as needed.
            this.resourcesTableAdapter.Fill(this.gantTest01DataSet.Resources);
            // TODO: This line of code loads data into the 'gantTest01DataSet.Appointments' table. You can move, or remove it, as needed.
            this.appointmentsTableAdapter.Fill(this.gantTest01DataSet.Appointments);
            
            this.appointmentsTableAdapter.Adapter.RowUpdated += new SqlRowUpdatedEventHandler(appointmentsTableAdapter_RowUpdated);

            this.schedulerControl1.OptionsCustomization.AllowAppointmentConflicts = AppointmentConflictsMode.Custom;
            this.schedulerControl1.AllowAppointmentConflicts+=new AppointmentConflictEventHandler(schedulerControl1_AllowAppointmentConflicts);
        }

        #region #enforcedependencies
        private void schedulerControl1_AllowAppointmentConflicts(object sender, AppointmentConflictEventArgs e)
        {
            e.Conflicts.Clear();

            AppointmentDependencyBaseCollection depCollectionDep = 
                schedulerDataStorage1.AppointmentDependencies.Items.GetDependenciesByDependentId(e.Appointment.Id);
            if (depCollectionDep.Count > 0) {
                if (CheckForInvalidDependenciesAsDependent(depCollectionDep, e.AppointmentClone))
                    e.Conflicts.Add(e.AppointmentClone);
            }

            AppointmentDependencyBaseCollection depCollectionPar = 
                schedulerDataStorage1.AppointmentDependencies.Items.GetDependenciesByParentId(e.Appointment.Id);
            if (depCollectionPar.Count > 0) {
                if (CheckForInvalidDependenciesAsParent(depCollectionPar, e.AppointmentClone))
                    e.Conflicts.Add(e.AppointmentClone);
            }
        }

        private bool CheckForInvalidDependenciesAsDependent(AppointmentDependencyBaseCollection depCollection, Appointment apt)
        {
            foreach (AppointmentDependency dep in depCollection) {
                if (dep.Type == AppointmentDependencyType.FinishToStart) {
                    DateTime checkTime = schedulerDataStorage1.Appointments.Items.GetAppointmentById(dep.ParentId).End;
                    if (apt.Start < checkTime)
                        return true;
                }
            }
            return false;
        }

        private bool CheckForInvalidDependenciesAsParent(AppointmentDependencyBaseCollection depCollection, Appointment apt)
        {
            foreach (AppointmentDependency dep in depCollection) {
                if (dep.Type == AppointmentDependencyType.FinishToStart) {
                    DateTime checkTime = schedulerDataStorage1.Appointments.Items.GetAppointmentById(dep.DependentId).Start;
                    if (apt.End > checkTime)
                        return true;
                }
            }
            return false;
        }
        #endregion #enforcedependencies


        #region #RowUpdatedHandlers
        int id = 0;
        private void appointmentsTableAdapter_RowUpdated(object sender, SqlRowUpdatedEventArgs e)
        {
            if (e.Status == UpdateStatus.Continue && e.StatementType == StatementType.Insert) {
                id = 0;
                using (SqlCommand cmd = new SqlCommand("SELECT @@IDENTITY", appointmentsTableAdapter.Connection)) {
                    id = Convert.ToInt32(cmd.ExecuteScalar());
                    e.Row["UniqueId"] = id;
                }
            }
        }
        #endregion #RowUpdatedHandlers

        #region #Appointment
        private void schedulerDataStorage1_AppointmentsChanged(object sender, PersistentObjectsEventArgs e)
        {
            CommitTask();
        }

        private void schedulerDataStorage1_AppointmentsDeleted(object sender, PersistentObjectsEventArgs e)
        {
            CommitTask();
        }
        private void schedulerDataStorage1_AppointmentsInserted(object sender, PersistentObjectsEventArgs e)
        {

            CommitTask();
            schedulerDataStorage1.SetAppointmentId(((Appointment)e.Objects[0]), id);
        }
        void CommitTask()
        {

            appointmentsTableAdapter.Update(gantTest01DataSet);
            this.gantTest01DataSet.AcceptChanges();
        }
        #endregion #Appointment
        #region #TaskDependencies
        private void schedulerDataStorage1_AppointmentDependenciesChanged(object sender, PersistentObjectsEventArgs e)
        {
            CommitTaskDependency();
        }

        private void schedulerDataStorage1_AppointmentDependenciesDeleted(object sender, PersistentObjectsEventArgs e)
        {
            CommitTaskDependency();
        }

        private void schedulerDataStorage1_AppointmentDependenciesInserted(object sender, PersistentObjectsEventArgs e)
        {
            CommitTaskDependency();
        }
        void CommitTaskDependency()
        {
            taskDependenciesTableAdapter.Update(this.gantTest01DataSet);
            this.gantTest01DataSet.AcceptChanges();
        }
        #endregion #TaskDependencies
    }
}

File: VB/GanttRestrictions/Form1.vb
Imports DevExpress.XtraScheduler
Imports System
Imports System.Data
Imports System.Data.SqlClient
Imports System.Windows.Forms

Namespace GanttRestrictions

    Public Partial Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            ' TODO: This line of code loads data into the 'gantTest01DataSet.TaskDependencies' table. You can move, or remove it, as needed.
            taskDependenciesTableAdapter.Fill(gantTest01DataSet.TaskDependencies)
            ' TODO: This line of code loads data into the 'gantTest01DataSet.Resources' table. You can move, or remove it, as needed.
            resourcesTableAdapter.Fill(gantTest01DataSet.Resources)
            ' TODO: This line of code loads data into the 'gantTest01DataSet.Appointments' table. You can move, or remove it, as needed.
            appointmentsTableAdapter.Fill(gantTest01DataSet.Appointments)
            AddHandler appointmentsTableAdapter.Adapter.RowUpdated, New SqlRowUpdatedEventHandler(AddressOf appointmentsTableAdapter_RowUpdated)
            schedulerControl1.OptionsCustomization.AllowAppointmentConflicts = AppointmentConflictsMode.Custom
            AddHandler schedulerControl1.AllowAppointmentConflicts, New AppointmentConflictEventHandler(AddressOf schedulerControl1_AllowAppointmentConflicts)
        End Sub

#Region "#enforcedependencies"
        Private Sub schedulerControl1_AllowAppointmentConflicts(ByVal sender As Object, ByVal e As AppointmentConflictEventArgs)
            e.Conflicts.Clear()
            Dim depCollectionDep As AppointmentDependencyBaseCollection = schedulerDataStorage1.AppointmentDependencies.Items.GetDependenciesByDependentId(e.Appointment.Id)
            If depCollectionDep.Count > 0 Then
                If CheckForInvalidDependenciesAsDependent(depCollectionDep, e.AppointmentClone) Then e.Conflicts.Add(e.AppointmentClone)
            End If

            Dim depCollectionPar As AppointmentDependencyBaseCollection = schedulerDataStorage1.AppointmentDependencies.Items.GetDependenciesByParentId(e.Appointment.Id)
            If depCollectionPar.Count > 0 Then
                If CheckForInvalidDependenciesAsParent(depCollectionPar, e.AppointmentClone) Then e.Conflicts.Add(e.AppointmentClone)
            End If
        End Sub

        Private Function CheckForInvalidDependenciesAsDependent(ByVal depCollection As AppointmentDependencyBaseCollection, ByVal apt As Appointment) As Boolean
            For Each dep As AppointmentDependency In depCollection
                If dep.Type = AppointmentDependencyType.FinishToStart Then
                    Dim checkTime As Date = schedulerDataStorage1.Appointments.Items.GetAppointmentById(dep.ParentId).End
                    If apt.Start < checkTime Then Return True
                End If
            Next

            Return False
        End Function

        Private Function CheckForInvalidDependenciesAsParent(ByVal depCollection As AppointmentDependencyBaseCollection, ByVal apt As Appointment) As Boolean
            For Each dep As AppointmentDependency In depCollection
                If dep.Type = AppointmentDependencyType.FinishToStart Then
                    Dim checkTime As Date = schedulerDataStorage1.Appointments.Items.GetAppointmentById(dep.DependentId).Start
                    If apt.End > checkTime Then Return True
                End If
            Next

            Return False
        End Function

#End Region  ' #enforcedependencies
#Region "#RowUpdatedHandlers"
        Private id As Integer = 0

        Private Sub appointmentsTableAdapter_RowUpdated(ByVal sender As Object, ByVal e As SqlRowUpdatedEventArgs)
            If e.Status = UpdateStatus.Continue AndAlso e.StatementType = StatementType.Insert Then
                id = 0
                Using cmd As SqlCommand = New SqlCommand("SELECT @@IDENTITY", appointmentsTableAdapter.Connection)
                    id = Convert.ToInt32(cmd.ExecuteScalar())
                    e.Row("UniqueId") = id
                End Using
            End If
        End Sub

#End Region  ' #RowUpdatedHandlers
#Region "#Appointment"
        Private Sub schedulerDataStorage1_AppointmentsChanged(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
            CommitTask()
        End Sub

        Private Sub schedulerDataStorage1_AppointmentsDeleted(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
            CommitTask()
        End Sub

        Private Sub schedulerDataStorage1_AppointmentsInserted(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
            CommitTask()
            schedulerDataStorage1.SetAppointmentId(CType(e.Objects(0), Appointment), id)
        End Sub

        Private Sub CommitTask()
            appointmentsTableAdapter.Update(gantTest01DataSet)
            gantTest01DataSet.AcceptChanges()
        End Sub

#End Region  ' #Appointment
#Region "#TaskDependencies"
        Private Sub schedulerDataStorage1_AppointmentDependenciesChanged(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
            CommitTaskDependency()
        End Sub

        Private Sub schedulerDataStorage1_AppointmentDependenciesDeleted(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
            CommitTaskDependency()
        End Sub

        Private Sub schedulerDataStorage1_AppointmentDependenciesInserted(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
            CommitTaskDependency()
        End Sub

        Private Sub CommitTaskDependency()
            taskDependenciesTableAdapter.Update(gantTest01DataSet)
            gantTest01DataSet.AcceptChanges()
        End Sub
#End Region  ' #TaskDependencies
    End Class
End Namespace
